src/ring_theory/derivation.lean

```diff
@@ -0,0 +1,210 @@
+  mul_smul := λ a1 a2 D, ext $ λ b, mul_smul _ _ _,
+  one_smul := λ D, ext $ λ b, one_smul A _,
+  smul_add := λ a D1 D2, ext $ λ b, smul_add _ _ _,
+  smul_zero := λ a, ext $ λ b, smul_zero _,
+  add_smul := λ a1 a2 D, ext $ λ b, add_smul _ _ _,
+  zero_smul := λ D, ext $ λ b, zero_smul A _ }
+
+@[simp] lemma smul_apply : (a • D) b = a • D b := rfl
+
+instance : is_scalar_tower R A (derivation R A M) :=
+⟨ λ x y z, ext (λ a, smul_assoc _ _ _) ⟩
+
+end
+
+section
+
+variables {R : Type*} [comm_ring R]
+variables {A : Type*} [comm_ring A] [algebra R A]
+
+section
+
+variables {M : Type*} [add_comm_group M] [module A M] [module R M] [is_scalar_tower R A M]
+variables (D : derivation R A M) {D1 D2 : derivation R A M} (r : R) (a b : A)
+
+@[simp] lemma map_neg : D (-a) = -D a := linear_map.map_neg D a
+@[simp] lemma map_sub : D (a - b) = D a - D b := linear_map.map_sub D a b
+
+instance : add_comm_group (derivation R A M) :=
+{ neg := λ D, ⟨-D, λ a b, by simp only [linear_map.neg_apply, smul_neg, neg_add_rev, leibniz,
+    linear_map.to_fun_eq_coe, coe_linear_map, add_comm]⟩,
+  add_left_neg := λ D, ext $ λ a, add_left_neg _,
+  ..derivation.add_comm_monoid }
+
+end
+
+section lie_structures
+
+/-! # Lie structures -/
+
+variables (D : derivation R A A) {D1 D2 : derivation R A A} (r : R) (a b : A)
+
+open ring_commutator
+
+/-- The commutator of derivations is again a derivation. -/
+def commutator (D1 D2 : derivation R A A) : derivation R A A :=
+⟨⁅D1, D2⁆, λ a b, by {simp only [commutator, map_add, id.smul_eq_mul, linear_map.mul_app,
+  leibniz, linear_map.to_fun_eq_coe, coe_linear_map, linear_map.sub_apply], ring }⟩
+
+instance : has_bracket (derivation R A A) := ⟨derivation.commutator⟩
+
+@[simp] lemma commutator_coe_linear_map : ↑⁅D1, D2⁆ = (⁅D1, D2⁆ : A →ₗ[R] A) := rfl
+
+lemma commutator_apply : ⁅D1, D2⁆ a = D1 (D2 a) - D2 (D1 a) := rfl
+
+instance : lie_ring (derivation R A A) :=
+{ add_lie := λ d e f, by { ext a, simp only [commutator_apply, add_apply, map_add], ring },
+  lie_add := λ d e f, by { ext a, simp only [commutator_apply, add_apply, map_add], ring },
+  lie_self := λ d, by { ext a, simp only [commutator_apply, add_apply, map_add], ring },
+  jacobi := λ d e f, by { ext a, simp only [commutator_apply, add_apply, map_sub], ring } }
+
+instance : lie_algebra R (derivation R A A) :=
+{ lie_smul := λ r d e, by { ext a, simp only [commutator_apply, map_smul, smul_sub, Rsmul_apply]},
+  ..derivation.Rsemimodule }
+
+end lie_structures
+
+end
+
+end derivation
+
+section comp_der
+
+namespace linear_map
+
+variables {R : Type*} [comm_semiring R]
+variables {A : Type*} [comm_semiring A] [algebra R A]
+variables {M : Type*} [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
+variables {N : Type*} [add_cancel_comm_monoid N] [semimodule A N] [semimodule R N]
+variables [is_scalar_tower R A M] [is_scalar_tower R A N]
+
+/-- The composition of a derivation and a linear map is a derivation. -/
```

Comment:
I should have added this together with the suggestion to swap the order of the arguments, sorry.
```suggestion
/-- The composition of a linear map and a derivation is a derivation. -/
```

Reply:
Ok thanks